postTransactionFee uses balanceTransaction

Unisay opened the pull request:
Rewrite the `postTransactionFee` endpoint so that:

- it uses `balanceTransaction` as its primary means of transaction balancing;
- it no longer directly depends upon the internal coin selection library.

### Issue Number

ADP-2269

Anviking committed:
Fix signing of external reward accounts

Unisay committed:
Extract tx fee calculation from delegation fee calculation

postTransactionFeeOld uses balanceTx

Anviking commented on lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs:
```diff
@@ -2344,51 +2346,44 @@ mkApiTransactionFromInfo ti wrk wid deposit info metadataSchema = do
-        let outs = addressAmountToTxOut <$> body ^. #payments
-        pp <- liftIO $ NW.currentProtocolParameters netLayer
-        let selectAssetsParams = W.SelectAssetsParams
-                { outputs = F.toList outs
```
Pretty sure we need to include the `outs` in the `PreSelection` inside `transactionFee`

Anviking commented on lib/wallet/src/Cardano/Wallet/Write/Tx/Balance.hs:
```diff
@@ -323,8 +323,7 @@ data UTxOAssumptions = forall k ktype. UTxOAssumptions
 -- depending on the 'k' of the supplied 'TransactionLayer'.
 allKeyPaymentCredentials
-    :: forall k. TransactionLayer k 'CredFromKeyK SealedTx
-    -> UTxOAssumptions
```
Can this be reverted? If you're calling `allKeyPaymentCredentials` then you should have `ktype ~ 'CredFromKeyK`. 

Secondly, I'd prefer to keep the existing formatting. While we do tend to write simple types on one line `foo :: Int -> Int`, I think it's in this case 1) a bit on the long side 2) breaks similarity to `allScriptPaymentCredentials`.

Unisay replied:
I guess its a matter of personal taste (for me this formatting is confusing as it has type-level and value-level mixed in one line).

Reverted.

Anviking replied:
Thanks.

> I guess its a matter of personal taste

While this is personal taste

>  I think it's in this case 1) a bit on the long side

this is not:

> 2) breaks similarity to allScriptPaymentCredentials.

— at least less directly so.

Anviking commented on lib/wallet/src/Cardano/Wallet.hs:
```diff
@@ -3841,3 +3878,12 @@ defaultChangeAddressGen arg proxy =
     ChangeAddressGen
         (genChange arg)
         (maxLengthAddressFor proxy)
+
```
```suggestion
-- | WARNING: Must never be used to create real transactions for submission to the blockchain as funds sent to a dummy change address would be irrecoverable.
```

Anviking commented on lib/wallet/src/Cardano/Wallet.hs:
```diff
@@ -2912,23 +2935,37 @@ delegationFee db@DBLayer{atomically, walletsDB} netLayer
+                    balanceTransaction @_ @_ @s
+                        nullTracer
+                        (Write.allKeyPaymentCredentials txLayer)
+                        (toBalanceTxPParams protocolParams)
```
Should be moved outside the loop

Anviking committed:
Adjust fee expectations in SHELLEY_TX_REDEEM_02

Anviking committed:
Add HasCallStack to test assertion helpers

To get better source code locations printed from failing integration
tests

jonathanknowles commented on lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs:
```diff
@@ -2344,52 +2346,70 @@ mkApiTransactionFromInfo ti wrk wid deposit info metadataSchema = do
+    -- transition to balanceTx they will temporarily compare a rewritten
+    -- postTransactionFee with an old postTransaction. They may also act as
+    -- golden tests in lieu of better/more ones on the unit level. A small
+    -- amount of padding is a cheap way to keep these checks.
```
Some typos fixed, and a slight rewording:

```suggestion
    -- 2. Our integration tests often assume that transaction fees are bounded
    -- absolutely by the estimate intervals provided by postTransactionFee.
    -- This assumption is flawed, since each estimate interval is merely a pair
    -- of 10- and 90-th percentile values: these are not absolute bounds.
    --
    -- But for the moment, it's still useful for integration tests to expect
    -- that eventual fees are bounded by prior fee estimates: during the risky
    -- transition to balanceTx they will compare the result of a rewritten
    -- version of postTransactionFee with the result of an unmodified version
    -- of postTransaction. They may also act as golden tests in lieu of
    -- better/more tests on the unit level. A small amount of padding is a
    -- cheap way to keep these checks.
```

jonathanknowles commented on lib/wallet/src/Cardano/Wallet.hs:
```diff
@@ -2270,40 +2276,47 @@ buildTransactionPure
+-- 'buildAndSignTransaction' achieves this by deriving the 'XPub' regardless
+-- of wallet type, from the root key. To avoid requiring another
+-- 'withRootKey' call, and to make the sketchy behaviour more explicit, we
+-- make 'buildAndSignTransactionPure' partial instead.
```
This hack existed before this PR, but perhaps we should create a technical debt ticket to record our intention to remove this hack?

If so, then perhaps we can add a `TODO` with a link to the ticket in the above comment?

Something like:
```hs
--- TODO: https://input-output.atlassian.net/browse/ADP-XXXX
--- The 'mkUnsignedTransaction' function takes a reward account ...
```

Anviking replied:
I thought I added one in another PR targeting this one, but here is a task at least: https://input-output.atlassian.net/browse/ADP-2933

jonathanknowles commented on lib/wallet/src/Cardano/Wallet.hs:
```diff
@@ -2270,40 +2276,47 @@ buildTransactionPure
+        Refl <- isShelleyKey
+        pure $ getRawKey $ Seq.rewardAccountKey @n @k walletState
+    where
+    isSeqState = testEquality (typeRep @s) (typeRep @(SeqState n k))
```
```suggestion
    fromMaybe notShelleyWallet $ do
        Refl <- isSeqState
        Refl <- isShelleyKey
        pure $ getRawKey $ Seq.rewardAccountKey @n @k walletState
  where
    isSeqState = testEquality (typeRep @s) (typeRep @(SeqState n k))
```

jonathanknowles commented on lib/wallet/src/Cardano/Wallet.hs:
```diff
@@ -2270,40 +2276,47 @@ buildTransactionPure
+        (error $ unwords
+            [ "toBalanceTxPParams: no nodePParams."
+            , "should only be possible in Byron, where"
+            , "withRecentEra should prevent this to be reached."
```
More fluent wording:
```suggestion
            , "This should only be possible in Byron, where withRecentEra"
            , "should prevent this from being reached."
```

jonathanknowles commented on lib/wallet/src/Cardano/Wallet.hs:
```diff
@@ -2854,28 +2867,7 @@ newtype Percentile (n :: Nat) a = Percentile a
-  where
-    throwInIO :: forall a. ExceptT ErrNoSuchWallet IO a -> IO a
-    throwInIO = runExceptT >=> either (throwIO . ExceptionNoSuchWallet) pure
+
```
```suggestion
```

jonathanknowles commented on lib/wallet/src/Cardano/Wallet.hs:
```diff
@@ -2887,21 +2879,57 @@ instance NFData DelegationFee
-    protocolParams <- liftIO $ currentProtocolParameters netLayer
+delegationFee
+    db@DBLayer {..} protocolParams txLayer ti era changeAddressGen walletId = do
+    --
```
Suggestion: delete this empty comment.
(Perhaps there used to be a comment on this line, but now it's just an empty comment.)

```suggestion
```

jonathanknowles commented on lib/wallet/src/Cardano/Wallet.hs:
```diff
@@ -2887,21 +2879,57 @@ instance NFData DelegationFee
+    -- Calculate the minimum deposit necessary if a given wallet wanted to
+    -- delegate to a pool. Said differently, this return either 0, or the value
+    -- of the key deposit protocol parameters if the wallet has no registered
+    -- stake key.
```
@Unisay Perhaps we could get away without this comment? Unless I'm mistaken, it doesn't tell us much beyond what the code is already telling us. (In other words, it seems like a natural language rewording of the case statement.)
```suggestion
```
What do you think?

jonathanknowles commented on lib/wallet/src/Cardano/Wallet.hs:
```diff
@@ -2982,11 +3032,32 @@ calculateFeePercentiles
             ) -> pure $ Fee requiredCost
         e -> throwE e
 
+-- | Make a pair of fee estimation percentiles more imprecise.
```
Suggestion: explain more precisely how this function makes the percentiles less precise.

```suggestion
-- | Make a pair of fee estimation percentiles more imprecise.
--
-- For a given number of bytes `n` and interval `(p, q)`, this function
-- computes the superinterval `(p - x, q + x)`, where `x` is the cost of
-- encoding `n` bytes according to the given protocol parameters.
```

jonathanknowles commented on lib/wallet/src/Cardano/Wallet.hs:
```diff
@@ -3841,3 +3912,12 @@ defaultChangeAddressGen arg proxy =
     ChangeAddressGen
         (genChange arg)
         (maxLengthAddressFor proxy)
+-- | WARNING: Must never be used to create real transactions for submission to the blockchain as funds sent to a dummy change address would be irrecoverable.
```
Request: wrap to fit line length limit.

```suggestion

-- WARNING: Must never be used to create real transactions for submission to the
-- blockchain as funds sent to a dummy change address would be irrecoverable.
```

jonathanknowles commented on lib/wallet/src/Cardano/Wallet.hs:
```diff
@@ -2982,11 +3032,32 @@ calculateFeePercentiles
+    coinDelta =
+        Coin.fromNatural . ceiling $ fromIntegral byteDelta * slope feeFunction
+
+    LinearFee feeFunction = pp ^. #txParameters . #getFeePolicy
```
Suggestion: explicitly specify the source and target type for `fromIntegral`.
Justification: `fromIntegral` is an unsafe function. If we do use it, then adding types can help readers to understand exactly what type conversions are happening, and judge whether the usage is correct.

```suggestion
      coinDelta :: Coin
      coinDelta
          = Coin.fromNatural
          . ceiling @Double @Natural
          $ fromIntegral @Word @Double byteDelta * slope

      LinearFee (LinearFunction {slope}) = pp ^. #txParameters . #getFeePolicy
```

jonathanknowles commented on lib/wallet/integration/src/Test/Integration/Framework/DSL.hs:
```diff
@@ -1729,12 +1729,13 @@ fundSharedWallet ctx amt sharedWals = do
        expectField
            (#balance . #available)
-           (`shouldBe` Quantity (faucetAmt - feeMax - amt)) ra
+           (`shouldBe` Quantity (faucetAmt - fee - amt)) ra
```
Perhaps this change was accidentally included in this PR?

On my machine, compilation of `Framework.DSL` fails with:
```
integration/src/Test/Integration/Framework/DSL.hs:1727:38: error: [-Wunused-matches, -Werror=unused-matches]
    Defined but not used: `feeMax'
     |
1727 |    (_, ApiFee (Quantity _) (Quantity feeMax) _ _) <- unsafeRequest ctx
```

jonathanknowles reviewed (commented):
Hi @Unisay 

I've left some suggestions and a couple of small requests!

Also, it seems this PR doesn't compile. (There's a compilation error in `Framework.DSL`). See: https://buildkite.com/input-output-hk/cardano-wallet/builds/24389#018751de-7ddb-4248-a4d2-98e5b78ab673/6-16

Unisay committed:
refactor: TRANSMETA_CREATE_01b

Anviking committed:
Add temporary workaround for BYRON_TRANS_CREATE_01

Anviking committed:
Fix TRANS_CREATE_01x

Anviking committed:
Include TTL when estimating tx fee

It is included in `postTransactionOld` so it seems like a good idea to
include it in `postTransactionFee`, even though it seems like the old
`postTransactionFee` implementation didn't.

Anviking committed:
Fix SHARED_WALLETS_DISCOVER_01

Anviking committed:
Add reference to ADP-2935

Unisay committed:
chore: use visible type applications for integral type conversions

Unisay committed:
Fix ApiError.json

Anviking committed:
Fix expectation for TRANSMETA_CREATE_03

Anviking committed:
fixup: compilation error and warning

Anviking committed:
fixup stylish

Unisay committed:
Remove unnecessary comment

Anviking commented on lib/wallet/src/Cardano/Wallet.hs:
```diff
@@ -2865,11 +2882,44 @@ delegationFee
+            $ throwInIO . mkNoSuchWalletError walletId
+            $ mapExceptT atomically (isStakeKeyRegistered walletId) <&> \case
+                False -> stakeKeyDeposit (fst protocolParams)
+                True -> Coin 0
```
👌 inlining seems fair
